packages/integration-sdk-entity-validator/src/validator.ts

```diff
@@ -0,0 +1,158 @@
+import Ajv, { AnySchema } from 'ajv';
+import addFormats from 'ajv-formats';
+import { addJ1Formats } from './j1Formats';
+import {
+  ajvErrorToEntityValidationError,
+  unknownPropertySymbol,
+  isEntityValidationError,
+  EntityValidationError,
+} from './entityValidationError';
+import { assertEntity } from './assertEntity';
+
+type ValidateEntityResult = {
+  isValid: boolean;
+  errors: EntityValidationError[] | null;
+  skippedSchemas: SkippedSchema[] | null;
+};
+
+type SkippedSchema = {
+  schemaId: string;
+  reason: 'not-found' | 'type-already-validated';
+  type: 'type' | 'class';
+};
+
+type ValidateEntityOptions = {
+  forceClassValidationWithValidatedType?: boolean;
+};
+
+const convertUnknownErrorToEntityValidationError = (
+  error: unknown,
+): EntityValidationError => {
+  if (isEntityValidationError(error)) {
+    return error;
+  }
+
+  if (error instanceof Error) {
+    return {
+      schemaId: 'unknown',
+      message: error.message,
+      property: unknownPropertySymbol,
+      validation: 'unknown',
+    };
+  }
+
+  return {
+    schemaId: 'unknown',
+    message: 'Unknown error',
+    property: unknownPropertySymbol,
+    validation: 'unknown',
+  };
+};
+
+export class EntityValidator {
+  private ajvInstance: Ajv;
+
+  constructor({ schemas }: { schemas?: AnySchema[] }) {
+    this.ajvInstance = addJ1Formats(
+      addFormats(
+        new Ajv({
+          strictSchema: false,
+          allErrors: true,
+        }),
+      ),
+    );
+
+    if (schemas) {
+      this.addSchemas(schemas);
+    }
+  }
+
+  addSchemas(schema: AnySchema | AnySchema[]) {
+    this.ajvInstance.addSchema(schema);
+    return this;
+  }
+
+  validateEntity(
+    entity: object,
+    {
+      forceClassValidationWithValidatedType = false,
+    }: ValidateEntityOptions = {},
+  ): ValidateEntityResult {
+    const errors: EntityValidationError[] = [];
+    const skippedSchemas: SkippedSchema[] = [];
+
+    try {
+      assertEntity(entity);
+
+      const classArray = Array.isArray(entity._class)
+        ? entity._class
+        : [entity._class];
+
+      const typeSchemaId = `#${entity._type}`;
+      const typeValidator = this.ajvInstance.getSchema(typeSchemaId);
+
+      if (typeValidator) {
+        const isValid = typeValidator(entity);
+
+        if (!isValid) {
+          errors.push(
+            ...(typeValidator.errors?.map((error) =>
+              ajvErrorToEntityValidationError(typeSchemaId, error),
+            ) ?? []),
+          );
+        }
+      } else {
+        skippedSchemas.push({
+          schemaId: typeSchemaId,
+          reason: 'not-found',
+          type: 'type',
+        });
+      }
+
+      if (!typeValidator || forceClassValidationWithValidatedType) {
```

Comment:
what's the use-case for `forceClassValidationWithValidatedType`?

Reply:
I'm not sure, that because it's false by default. But you may want to validate the latest class schema as well cause the type schema has a outdated class schema merged in it.

Reply:
I was under the impression that when a class schema is updated, the type schemas that used it would also be updated (and published). But I also don't really know how that would work tbh.

Reply:
That's the idea, but I'm not sure when we are getting there. So I think having this flag available might be helpful for now. In the best scenario it won't be used and we can remove it.